coderd/notifications/reports/generator.go

```diff
@@ -98,73 +96,87 @@ func (i *reportGenerator) Close() error {
 	return nil
 }
 
-const failedWorkspaceBuildsReportFrequencyDays = 7
+const (
+	failedWorkspaceBuildsReportFrequency      = 7 * 24 * time.Hour
+	failedWorkspaceBuildsReportFrequencyLabel = "week"
+)
 
 func reportFailedWorkspaceBuilds(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) error {
 	now := clk.Now()
-	since := now.Add(-failedWorkspaceBuildsReportFrequencyDays * 24 * time.Hour)
+	since := now.Add(-failedWorkspaceBuildsReportFrequency)
+
+	// Firstly, check if this is the first run of the job ever
+	reportLog, err := db.GetNotificationReportGeneratorLogByTemplate(ctx, notifications.TemplateWorkspaceBuildsFailedReport)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return xerrors.Errorf("unable to read report generator log: %w", err)
+	}
+	if xerrors.Is(err, sql.ErrNoRows) {
+		// First run? Check-in the job, and get back after one week.
+		logger.Info(ctx, "report generator is executing the job for the first time", slog.F("notification_template_id", notifications.TemplateWorkspaceBuildsFailedReport))
 
-	statsRows, err := db.GetWorkspaceBuildStatsByTemplates(ctx, dbtime.Time(since).UTC())
+		err = db.UpsertNotificationReportGeneratorLog(ctx, database.UpsertNotificationReportGeneratorLogParams{
+			NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			LastGeneratedAt:        dbtime.Time(now).UTC(),
+		})
+		if err != nil {
+			return xerrors.Errorf("unable to update report generator logs (first time execution): %w", err)
+		}
+		return nil
+	}
+
+	// Secondly, check if the job has not been running recently
+	if !reportLog.LastGeneratedAt.IsZero() && reportLog.LastGeneratedAt.Add(failedWorkspaceBuildsReportFrequency).After(now) {
+		return nil // reports sent recently, no need to send them now
+	}
+
+	// Thirdly, fetch workspace build stats by templates
+	templateStatsRows, err := db.GetWorkspaceBuildStatsByTemplates(ctx, dbtime.Time(since).UTC())
 	if err != nil {
 		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
 	}
-	sort.Slice(statsRows, func(i, j int) bool {
-		return statsRows[i].TemplateName < statsRows[j].TemplateName
-	})
 
-	reportGeneratedNow := map[uuid.UUID]bool{}
-	for _, stats := range statsRows {
-		var failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow
-		reportData := map[string]any{}
-
-		if stats.FailedBuilds > 0 {
-			failedBuilds, err = db.GetFailedWorkspaceBuildsByTemplateID(ctx, database.GetFailedWorkspaceBuildsByTemplateIDParams{
-				TemplateID: stats.TemplateID,
-				Since:      dbtime.Time(since).UTC(),
-			})
-			if err != nil {
-				logger.Error(ctx, "unable to fetch failed workspace builds", slog.F("template_id", stats.TemplateID), slog.Error(err))
-				continue
-			}
+	for _, stats := range templateStatsRows {
+		if xerrors.Is(ctx.Err(), context.Canceled) {
+			logger.Debug(ctx, "context is canceled, quitting")
+			break
+		}
```

Comment:
This is the idiomatic way to express this.

```suggestion
		select {
			case <-ctx.Done():
				logger.Debug(ctx, "context is canceled, quitting", slog.Error(ctx.Err())
				break
			default:				
		}
```